Bug 1548545 - gpg sign the artifacts we publish to maven.mozilla.org via github releases

MihaiTabara opened the pull request:
This PR does the following:

1. Beetmover no longer downloads + explode`maven.zip` but instead it reads each individual file via declarative artifacts.
2. Adds GPG signing for all individual artifacts in the github release. Because doing the same for snapshot releases is slightly more complex, we're rolling out support for just github releases for now. The snapshots countepart are tracked in [bug 1558795](https://bugzilla.mozilla.org/show_bug.cgi?id=1558795). For this reason there's some function duplications but we expect to have those removed early in Q3 when we add support for snapshots as well. 

@JohanLorenzo 
* PR green graph is [here](https://tools.taskcluster.net/groups/V7nmTPRyT0yi17RYIgCegg).
* staging snapshot green releas is [here](https://tools.taskcluster.net/groups/Z47_gR8nR0OoCKyU98aDgQ). Artifacts pushed to maven S3 too, see example [here](https://maven-snapshots.stage.mozaws.net/?prefix=maven2/org/mozilla/components/browser-state/3.0.0-SNAPSHOT/).
* staging github green release is [here](https://tools.taskcluster.net/groups/Cc-FY54ASAORsbUU1cFXHQ). Artifacts pushed to maven S3 too, see example [here](https://maven-default.stage.mozaws.net/?prefix=maven2/org/mozilla/components/service-telemetry/3.0.0/).

I still don't understand why TC bot says it can't submit the graph, even though it actually does that. Still debugging that.

MihaiTabara commented on automation/taskcluster/lib/tasks.py:
```diff
@@ -144,33 +179,52 @@ def craft_beetmover_task(
             "releaseProperties": {
                 "appName": "snapshot_components" if is_snapshot else "components",
             },
-            "version": "{}-SNAPSHOT".format(version) if is_snapshot else version,
```
Just realized this might break our snapshots. I'll check against staging snapshots to ensure this still works.

MihaiTabara replied:
This indeed broke snapshots as that's still expected. Working on a fix.

MihaiTabara commented on automation/taskcluster/lib/tasks.py:
```diff
@@ -86,6 +97,29 @@ def craft_compare_locales_task(self):
+            dependencies=[build_task_id],
+            routes=[],
+            scopes=[
+                "project:mobile:android-components:releng:signing:cert:{}-signing".format("dep" if is_staging else "release"),
```
@JohanLorenzo You were right, I double-checked too in https://github.com/mozilla-releng/build-puppet/blob/master/modules/signing_scriptworker/templates/passwords.json.erb#L23 vs https://github.com/mozilla-releng/build-puppet/blob/master/modules/signing_scriptworker/templates/passwords.json.erb#L47 and they are the same. Rather than adding a new layer of confusion, I'd say we keep `release-signing` for snapshots as well. What do you think?

JohanLorenzo replied:
Fine by me!

MihaiTabara commented on .taskcluster.yml:
```diff
@@ -213,34 +213,33 @@ tasks:
                       - scopes:
                           # XXX Replace with your own hook, if you perform staging releases
-                          - assume:hook-id:project-mobile/android-components-snapshot-release
+                          # FIXME: temporarily to test this in staging snapshot releases
```
Note to self: I still need to revert this to production.

MihaiTabara commented on automation/taskcluster/lib/tasks.py:
```diff
@@ -49,18 +64,24 @@ def craft_build_task(self, module_name, gradle_tasks, subtitle='', run_coverage=
         )
 
-    def craft_wait_on_builds_task(self, dependencies):
+    def craft_barrier_task(self, dependencies):
```
I renamed this to a more generic name as I didn't want to duplicate it. Snapshot beetmover needs to wait on all builds tasks, while github release's beetmover need to wait for all signing tasks. I believe barrier is a better name regardless as we might need to add more tasks in the future at some point.

mitchhentges replied:
Why do we want to wait for all signing tasks before we do pushes?
I'm :+1: on waiting for builds, since it's possible for a subset of modules to fail to compile.
However, for beetmoving, can't we push as soon as signing for a module finishes? Is there a possible scenario for a subset of signing jobs to fail?

MihaiTabara replied:
This is a very good point. I've amended the signing tasks to depend on the barrier instead. Let's parallelize signing and beetmover after barrier upon builds is completed 👍

MihaiTabara commented on automation/taskcluster/lib/tasks.py:
```diff
@@ -127,7 +172,78 @@ def craft_ui_tests_task(self):
+
+    # XXX: temporarily until we add signing support in snapshots in bug 155879
+    # in early Q3
+    def craft_snapshot_beetmover_task(
```
I forked and kept the old implementation for as long as we have snapshots still un-signed. Once that support is added, I'll attempt to meld the two in the same function as it currently is. At this point they were far too different to mess around the functions so I preffered this way. It's temporarily anyway.

MihaiTabara commented on automation/taskcluster/decision_task.py:
```diff
@@ -218,39 +219,111 @@ def _get_release_gradle_tasks(module_name, is_snapshot):
-    version = components_version() if version is None else version
+# XXX: TO DELETE once bug 1558795 is fixed in early Q3
+def release_snapshot(components, is_snapshot, is_staging):
+    version = components_version()
```
I forked this from default `release()` so that we don't have to add too many `if snapshot do this else do that` kind of statements. We're adding snapshot support soon anyway so these two functions will be merged into a single one anyway.

MihaiTabara commented on automation/taskcluster/lib/tasks.py:
```diff
@@ -86,6 +100,30 @@ def craft_compare_locales_task(self):
 
+    def craft_sign_task(self, build_task_id, artifacts, component_name, is_staging):
+        payload = {
+            "maxRunTime": 600,
```
Good point, I removed it from all scriptworker tasks. I left it only for build tasks that run in docker-workers.

MihaiTabara commented on automation/taskcluster/lib/tasks.py:
```diff
@@ -127,7 +165,77 @@ def craft_ui_tests_task(self):
+    ):
+        if is_snapshot:
+            if is_staging:
+                bucket_name = 'maven-snapshot-staging'
```
@JohanLorenzo Hm, yes and no. It's currently used in the `craft_beetmover_snapshot_task` but that function is going away in the next two weeks along with the support for GPG we're adding in snapshots. Eventually, that function will be retired as the https://github.com/mozilla-mobile/android-components/pull/3084/files#diff-59f90d509bc6ae2c9d877250c5ba49c5R235 depicts and will mold it back in this one. If you have strong feelings towards it, I can definitely move them into a separate function, but it's restricted to beetmover only.

MihaiTabara commented:
Addressed all reviewing comments from @JohanLorenzo 

Re-testing things:
* PR - https://tools.taskcluster.net/groups/dHA25N3URaKrreaUXzrVFw
* Staging release - https://tools.taskcluster.net/groups/YyQwihouTI2fzsXz41uWpQ
* Staging snapshot release - https://tools.taskcluster.net/groups/RdZoq0UwTzuRkK1jYgmJKQ/tasks/RdZoq0UwTzuRkK1jYgmJKQ/details

MihaiTabara commented:
Green graphs everywhere. @pocmo mind if you glance/stamp this too so that we can merge? thanks!
* PR - https://tools.taskcluster.net/groups/TMfVwprDQ22UXY3Jla5taA
* Staging release - https://tools.taskcluster.net/groups/YyQwihouTI2fzsXz41uWpQ
* Staging snapshot release - https://tools.taskcluster.net/groups/RXQIuN4zT-KwobibTjJBEw
